indexd/index/__init__.py

```diff
@@ -0,0 +1,22 @@
+from functools import wraps
+from flask import request
+
+from indexd.errors import UserError
+
+
+def request_args_to_params(func):
```

Comment:
Looking at this again I'm wondering why not just  use `flask.request.args`?

Reply:
Mainly cos of clarity in method parameters. I still use `flask.request.args`, but just in a single place

Reply:
The rest of the code already raises `NotImplementedError` for abstract methods, yet :). For this endpoints, Performance hit? I am not sure about that, but doing it this way, I can easily report parameter errors to back to the client without have to write much code, which I think is one of the benefits of python

Reply:
I guess the other code is a bad example. :slightly_smiling_face: 

Copying the values out from `flask.request.args` into another dict (the `kwargs`) is strictly speaking more costly than just using it directly, as is passing in extra arguments to the route functions. It's not a _dramatic_ performance hit, but I bring it up since you seemed concerned about the cost for a dictionary conversion from the URLs driver functions elsewhere. Getting the arguments out is still easy enough to do inline and if you want to actually return which parameters were unexpected you can't really avoid handling that separately.

Reply:
I would suggest instead of messing with the route functions to just define separate functions that actually handle the computations, and in the routes just do this
```
return flask.jsonify(driver.query_urls(flask.request.args))
# let the function raise UserError or whatever, which gets handled automatically
```

Reply:
To be sure I follow, use of python's standard concept of decorators is expensive and possibly confusing, but I can replace the decorator with a method that does the exact same thing and add some code to detect invalid parameters?

Reply:
The sample you provided will raise an unexpected argument exception once there is a typo, which will skip any exception handling within the methods. I am also not worried about the performance here cos I know its just input parameters (much less in number) and I am not potentially looping over thousands of record as would be the case when manipulating the result.

Reply:
If you can, take a closer look at what using the decorator approach provides. Only unknown named parameters (probably typos or intentional invalid params) gets passed as **kwargs, and I am very sure it does not interfere with flask internal route system.

Reply:
you can use the same `**kwargs` method as now if you just pass `flask.request.args` to another function. the only difference is you don't need to put a decorator on the route functions to pass in the parameters as arguments.

Reply:
suppose you change `query_metadata_by_key` to take a `**kwargs`, move the handling to there and raise `UserError` on unexpected arguments, and you can call `record_list = blueprint.driver.query_metadata_by_key(flask.request.args)`. does that make sense?